[Merged by Bors] - feat(data/nat/prime): prime divisors

stuart-presnell opened the pull request:
Adding some basic lemmas about `factors` and `factors.to_finset`

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

stuart-presnell committed:
feat(number_theory/divisors): prime divisors

stuart-presnell committed:
Eliminating `finish`es and non-terminal `simp`s

eric-wieser commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +403,82 @@ begin
+lemma zero_no_prime_divisors : prime_divisors 0  = (∅:finset ℕ) :=
+by { unfold prime_divisors divisors, simp }
+
+lemma prime_divisors_eq_prime_factors (n:ℕ) : n.prime_divisors = n.factors.to_finset :=
```
```suggestion
lemma prime_divisors_eq_to_finset_factors (n:ℕ) : n.prime_divisors = n.factors.to_finset :=
```

eric-wieser commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +403,82 @@ begin
+end
+
+/-- 1 has no prime divisors -/
+lemma one_no_prime_divisors : prime_divisors 1 = (∅:finset ℕ) :=
```
```suggestion
lemma prime_divisors_one : prime_divisors 1 = (∅ : finset ℕ) :=
```
Is more in line with mathlib naming

stuart-presnell committed:
Update src/number_theory/divisors.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

stuart-presnell committed:
Renamed some lemmas after reviewer comment

stuart-presnell committed:
Renamed some lemmas after reviewer comment (fixed)

eric-wieser commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +403,82 @@ begin
+  by_cases hb0 : b = 0,
+    { rw [hb0, coprime_zero_right] at h_ab_coprime,
+      rw [hb0, h_ab_coprime, mul_zero, prime_divisors_zero, prime_divisors_one], simp },
+  have hab : a * b ≠ 0, { exact mul_ne_zero ha0 hb0 },
```
You should extract everything from this point onwards to a standalone `prime_divisors_mul_of_ne_zero` lemma.

eric-wieser commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +403,82 @@ begin
+end
+
+/-- For coprime `a` and `b`, prime divisors of `(a * b)` are the union of those of `a` and `b` -/
+lemma coprime_prime_divisors_product {a b : ℕ} (h_ab_coprime : coprime a b) :
```
```suggestion
lemma prime_divisors_mul_of_coprime {a b : ℕ} (h_ab_coprime : coprime a b) :
```
`*` is called `mul` not `product` according to the naming guide

stuart-presnell committed:
Extracted `prime_divisors_mul_of_ne_zero` per reviewer suggestion

Ruben-VandeVelde commented on src/data/nat/gcd.lean:
```diff
@@ -453,4 +453,15 @@ begin
+  rw coprime_iff_gcd_eq_one at h_ab_coprime,
+  have h1 := dvd_gcd hka hkb,
+  rw h_ab_coprime at h1,
+  exact (nat.dvd_one).mp h1,
```
```suggestion
  exact nat.dvd_one.mp h1,
```

Ruben-VandeVelde commented on src/number_theory/divisors.lean:
```diff
@@ -47,6 +48,10 @@ def proper_divisors : finset ℕ := finset.filter (λ x : ℕ, x ∣ n) (finset.
 ((finset.Ico 1 (n + 1)).product (finset.Ico 1 (n + 1))).filter (λ x, x.fst * x.snd = n)
 
+/-- `prime_divisors n` is the `finset` of prime natural numbers that divide `n`. -/
+def prime_divisors (n:ℕ) := (divisors n).filter prime
```
```suggestion
def prime_divisors := (divisors n).filter prime
```

(There's a `variable` statement that provides it above)

eric-wieser commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +403,90 @@ begin
+end
+
+/-- If `a * b ≠ 0` then the prime divisors of `(a * b)` are the union of those of `a` and `b` -/
+lemma prime_divisors_mul_of_ne_zero {a b : ℕ} (hab : a * b ≠ 0) :
```
Oh, I meant
```suggestion
lemma prime_divisors_mul_of_ne_zero {a b : ℕ} (ha : a ≠ 0) (hb : b ≠ 0) :
```

Ruben-VandeVelde commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +403,90 @@ begin
 
+
+
+-- # Prime divisors
```
Should be a docstring

Ruben-VandeVelde commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +403,90 @@ begin
+
+lemma prime_divisors_mem {n p : ℕ} (hn : n ≠ 0) : p ∈ n.prime_divisors ↔ p ∣ n ∧ prime p :=
+by { unfold prime_divisors, rw [mem_filter, mem_divisors],
+    rw [and.congr_left_iff, and_iff_left_iff_imp], intros, exact hn }
```
```suggestion
by { unfold prime_divisors, rw [mem_filter, mem_divisors, and_iff_left hn] }
```

Ruben-VandeVelde commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +403,90 @@ begin
+
+/-- 1 has no prime divisors -/
+lemma prime_divisors_one : prime_divisors 1 = (∅ : finset ℕ) :=
+by { apply filter_singleton prime }
```
```suggestion
filter_singleton prime 1
```

stuart-presnell committed:
Amending `prime_divisors_mul_of_ne_zero` per reviewer suggestion

Ruben-VandeVelde commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +403,89 @@ begin
+lemma prime_divisors_mem {n p : ℕ} (hn : n ≠ 0) : p ∈ n.prime_divisors ↔ p ∣ n ∧ prime p :=
+by { unfold prime_divisors, rw [mem_filter, mem_divisors],
+    rw [and.congr_left_iff, and_iff_left_iff_imp], intros, exact hn }
+
```
I'd also add:
```suggestion

lemma prime_of_mem_prime_divisors {n p : ℕ} (h : p ∈ n.prime_divisors) : prime p :=
by { rw [prime_divisors, mem_filter] at h, exact h.2 }

lemma dvd_of_mem_prime_divisors {n p : ℕ} (h : p ∈ n.prime_divisors) : p ∣ n :=
by { rw [prime_divisors, mem_filter] at h, exact dvd_of_mem_divisors h.1 }

```

Ruben-VandeVelde commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +403,89 @@ begin
+  by_contra H, rw H at hx,
+  suffices : y = 1, { rw this at hy2, exact not_prime_one hy2 },
+  exact eq_one_of_dvd_coprimes h_ab_coprime hx.1 hy1,
+end
```
With these extra lemmas, you don't need to case on 0:
```suggestion
lemma coprime_prime_divisors_disjoint {a b : ℕ} (hab : coprime a b) :
  disjoint a.prime_divisors b.prime_divisors :=
begin
  rw disjoint_iff_ne,
  rintros x ha - hb rfl,
  apply not_prime_one,
  rw ←eq_one_of_dvd_coprimes hab (dvd_of_mem_prime_divisors ha) (dvd_of_mem_prime_divisors hb),
  exact prime_of_mem_prime_divisors ha,
end
```

stuart-presnell replied:
Very nice! Thanks!

Ruben-VandeVelde commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +403,89 @@ begin
+lemma prime_divisors_mul_of_ne_zero {a b : ℕ} (ha : a ≠ 0) (hb : b ≠ 0) :
+  (a * b).prime_divisors = a.prime_divisors ∪ b.prime_divisors :=
+begin
+  have hab : a * b ≠ 0, { exact mul_ne_zero ha hb },
```
```suggestion
  have hab : a * b ≠ 0 := mul_ne_zero ha hb,
```

Ruben-VandeVelde commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +403,89 @@ begin
+begin
+  by_cases ha : a = 0,
+    { rw [ha, coprime_zero_left] at h_ab_coprime,
+      rw [ha, h_ab_coprime, zero_mul, prime_divisors_zero, prime_divisors_one], simp },
```
```suggestion
  { rw [ha, coprime_zero_left] at h_ab_coprime,
    rw [ha, h_ab_coprime, zero_mul, prime_divisors_zero, prime_divisors_one, finset.empty_union] },
```

Ruben-VandeVelde commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +403,89 @@ begin
+      rw [ha, h_ab_coprime, zero_mul, prime_divisors_zero, prime_divisors_one], simp },
+  by_cases hb : b = 0,
+    { rw [hb, coprime_zero_right] at h_ab_coprime,
+      rw [hb, h_ab_coprime, mul_zero, prime_divisors_zero, prime_divisors_one], simp },
```
```suggestion
  { rw [hb, coprime_zero_right] at h_ab_coprime,
    rw [hb, h_ab_coprime, mul_zero, prime_divisors_zero, prime_divisors_one, finset.empty_union] },
```

Ruben-VandeVelde commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +403,89 @@ begin
+/-- The only prime divisor of positive prime power `p^k` is `p` itself -/
+lemma prime_pow_prime_divisor {p k : ℕ} (hk : k ≠ 0) (hp: prime p) : prime_divisors (p^k) = {p} :=
+begin
+  have h1 : (p^k) ≠ 0, { exact pow_ne_zero k (prime.ne_zero hp) },
```
```suggestion
  have h1 : (p^k) ≠ 0 := pow_ne_zero k hp.ne_zero,
```

Ruben-VandeVelde commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +403,89 @@ begin
+  { rintros ⟨hq1, hq2⟩,
+    rw ←(prime_dvd_prime_iff_eq hq2 hp),
+    exact prime.dvd_of_dvd_pow hq2 hq1 },
+  { intros h, rw h, use [dvd_pow_self p hk, hp] },
```
```suggestion
  { rintros rfl, exact ⟨dvd_pow_self q hk, hp⟩ },
```

Ruben-VandeVelde commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +403,89 @@ begin
+by { rw [←(pow_one p), (prime_pow_prime_divisor one_ne_zero hp), pow_one] }
+
+
+
```
```suggestion
```

Ruben-VandeVelde commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +403,89 @@ begin
+
+/-- The only prime divisor of prime `p` is `p` itself -/
+lemma prime_prime_divisor {p:ℕ} (hp: prime p) : prime_divisors p = {p} :=
+by { rw [←(pow_one p), (prime_pow_prime_divisor one_ne_zero hp), pow_one] }
```
```suggestion
lemma prime_prime_divisor {p : ℕ} (hp: prime p) : prime_divisors p = {p} :=
by { rw [←pow_one p, prime_pow_prime_divisor one_ne_zero hp, pow_one] }
```

Ruben-VandeVelde reviewed (commented):
Nice! Some suggestions below

stuart-presnell committed:
Some new lemmas and many simplifications from reviewer comments

ericrbg commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +403,86 @@ begin
+
+/-- # Prime divisors -/
+
+lemma prime_divisors_mem {n p : ℕ} (hn : n ≠ 0) : p ∈ n.prime_divisors ↔ p ∣ n ∧ prime p :=
```
I think mathlib naming usually would say `mem_prime_divisors`.

ericrbg commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +403,86 @@ begin
+/-- # Prime divisors -/
+
+lemma prime_divisors_mem {n p : ℕ} (hn : n ≠ 0) : p ∈ n.prime_divisors ↔ p ∣ n ∧ prime p :=
+by { unfold prime_divisors, rw [mem_filter, mem_divisors, and_iff_left hn] }
```
```suggestion
by rw [prime_divisors, mem_filter, mem_divisors, and_iff_left hn]
```
should work too

ericrbg commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +403,86 @@ begin
+lemma prime_divisors_one : prime_divisors 1 = (∅ : finset ℕ) :=
+filter_singleton prime 1
+
+lemma prime_divisors_eq_to_finset_factors (n : ℕ) : n.prime_divisors = n.factors.to_finset :=
```
if `factors` exists, why do we need this? is it for those "setty" lemmas? why not just make an api for `n.factors.to_finset`?

eric-wieser replied:
Yeah, this crossed my mind too. Certainly we need a lemma that says `n.factors.to_finset = (finset.range (n+1)).filter prime`, but I don't know if we actually need a special name for this operation

stuart-presnell replied:
Mainly I proved `prime_divisors_eq_to_finset_factors` just because Ruben Van de Velde asked if they were equal [in the Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/Proof.20of.20Euler's.20product.20formula.20for.20totient/near/261380440).  I'm not using this lemma for anything at the moment. 

I wondered in that thread whether it would be better to change the definition of `prime_divisors` to `factors.to_finset`, but I didn't want to make the change just speculatively and then have to tidy up.  But if this would be a better definition — if it's more efficient or in some other way substantially nicer than `(divisors n).filter prime` — then I'm happy to change it.

stuart-presnell replied:
So would you suggest just taking out `prime_divisors_eq_to_finset_factors`?

stuart-presnell replied:
Ok, I've taken out `prime_divisors_eq_to_finset_factors`, which I wasn't using, and replaced it with the following, which I've found useful in some other work elsewhere:
```
lemma factor_of_prime_divisor {p n : ℕ} (hp : p ∈ n.prime_divisors) : p ∈ n.factors
```

eric-wieser replied:
Sorry, I don't think either of us were asking for you to remove the lemma; rather, the lemma was the hint to us that `prime_divisors` _maybe_ shouldn't exist at all.

stuart-presnell replied:
Ah, I see, because the lemma reveals that it's essentially just an alias for `factors.to_finset`, and so the properties satisfied by `prime_divisors` should follow straightforwardly from the corresponding properties of `factors`.

stuart-presnell replied:
Coming back to this: when @ericrbg says "why not just make an api for `n.factors.to_finset`?" what exactly do you mean?  (Sorry, still quite new to this!)  Do you mean re-writing the lemmas and proofs in this PR to use `factors.to_finset` instead of `prime_divisors` (and remove the definition of the latter)?  Or would you expect that essentially everything in this PR would end up being trivial after such a translation and so not worth having?  I'm happy to rewrite this if it's worth having, and happy to drop it if it's not.

ericrbg replied:
precisely, just rewrite the lemmas that don't already exist for `n.factors`/`n.factors.to_finset` instead of `prime_divisors`.  I think everything up to `coprime_prime_divisors_disjoint` should even generalize to the list setting.

stuart-presnell replied:
Great, thanks

stuart-presnell committed:
Further amendments after reviewer suggestions

stuart-presnell committed:
Replace `prime_divisors_eq_to_finset_factors` with `factor_of_prime_divisor`

stuart-presnell committed:
Major re-write to eliminate `prime_divisors` terminology

I've re-written all the material in this PR to use `factors.to_finset` instead of `prime_divisors`.  Some of the lemmas have been re-expressed as lemmas about `factors`.  Some were revealed to be trivial after this re-write and have been discarded.

eric-wieser commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +400,77 @@ begin
+  by_cases hn : n = 0,
+  { convert dvd_zero p },
+  { rwa ←mem_factors_iff_dvd (pos_iff_ne_zero.mpr hn) (prime_of_mem_factors h) }
+end
```
This lemma should go right next to `nat.mem_factors_iff_dvd` in `data/nat/prime.lean`

eric-wieser commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +400,77 @@ begin
+  split,
+  { rintros ⟨hq1, hq2⟩, exact (prime_dvd_prime_iff_eq hq1 hp).mp (prime.dvd_of_dvd_pow hq1 hq2) },
+  { intros h, rw h, use [hp, dvd_pow_self p hk] },
+end
```
This feels like it should much easier if you start with `rw hp.factors_pow`

stuart-presnell replied:
Ah yes, much better.  The main body of the proof becomes `(list.repeat p k).to_finset = {p}`, which ought to exist as a general lemma about `lists` but apparently doesn't.  So I've added this lemma to `data/finset/basic` and then applied it here.

eric-wieser commented on src/number_theory/divisors.lean:
```diff
@@ -398,4 +400,77 @@ begin
+end
+
+lemma prime_divisors_eq_to_filter_divisors_prime (n : ℕ) :
+n.factors.to_finset = (divisors n).filter prime
```
I think all the lemmas except this one probably belong in `data/nat/prime`, since they're all about `nat.factors` not `divisors`.

stuart-presnell replied:
Yeah, I wondered whether they belonged somewhere else now but I wasn't sure where.  I'll move them over.

stuart-presnell committed:
Adding `repeat_to_finset_pos` to simplify a proof

stuart-presnell committed:
Moving material from `number_theory/divisors` to `data/nat/prime`

stuart-presnell committed:
Removed name from `Authors`, and doc-string on `prime_divisors`

eric-wieser commented on src/data/finset/basic.lean:
```diff
@@ -1823,6 +1823,19 @@ end
 to_finset_eq_of_perm _ _ (reverse_perm l)
 
+lemma repeat_to_finset_zero {a : α} : (repeat a 0).to_finset = (∅:finset α)
+  := by rw [list.repeat.equations._eqn_1 a, list.to_finset_nil]
```
I don't think we need this lemma

eric-wieser commented on src/data/finset/basic.lean:
```diff
@@ -1823,6 +1823,19 @@ end
+
+lemma repeat_to_finset_pos {a : α} {n : ℕ} (hn : 0 < n):
+  (list.repeat a n).to_finset = {a}
+  :=
```
```suggestion
lemma to_finset_repeat_of_pos (a : α) {n : ℕ} (hn : 0 < n) :
  (list.repeat a n).to_finset = {a} :=
```
This is about `to_finset (list.repeat a n)`, so `to_finset` should appear first in the name like it does above.

eric-wieser commented on src/data/nat/prime.lean:
```diff
@@ -633,6 +633,14 @@ lemma mem_factors_iff_dvd {n p : ℕ} (hn : 0 < n) (hp : prime p) : p ∈ factor
  λ h, mem_list_primes_of_dvd_prod hp (@prime_of_mem_factors n) ((prod_factors hn).symm ▸ h)⟩
 
+lemma dvd_of_mem_factors {n p : ℕ} (h : p ∈ n.factors) : p ∣ n
+  :=
```
Mathlib style is:
```suggestion
lemma dvd_of_mem_factors {n p : ℕ} (h : p ∈ n.factors) : p ∣ n :=
```

eric-wieser replied:
(this applies throughout)

stuart-presnell replied:
Ah, sorry, I knew this and failed to spot it.  I've gone through and fixed this throughout.

stuart-presnell committed:
Fixed formatting errors per reviewer comment

stuart-presnell committed:
Merge branch 'master' into SP_prime_divisors

ericrbg commented on src/data/finset/basic.lean:
```diff
@@ -1823,6 +1823,14 @@ end
+begin
+  ext x,
+  rw [list.mem_to_finset, finset.mem_singleton, list.mem_repeat, and_iff_right_iff_imp],
+  exact imp_intro (ne_of_gt hn),
```
I'd suggest passing in `n ≠ 0` as `hn` here, and then you can replace both lines with `simp [hn, list.mem_repeat]` if you want; there's dot notation to get from one to another though (  `‹n ≠ 0›.bot_lt = 0 < n` and `‹0 < n›.ne' = n ≠ 0`).

stuart-presnell replied:
Oh that's great, thanks!  I'm never quite sure whether to use `n ≠ 0` or `0 < n` as an argument, but those two simple conversions will make things much simpler.  I wonder why `library_search` doesn't bring them up.

ericrbg commented on src/data/nat/prime.lean:
```diff
@@ -633,6 +633,13 @@ lemma mem_factors_iff_dvd {n p : ℕ} (hn : 0 < n) (hp : prime p) : p ∈ factor
+begin
+  by_cases hn : n = 0,
+  { convert dvd_zero p },
+  { rwa ←mem_factors_iff_dvd (pos_iff_ne_zero.mpr hn) (prime_of_mem_factors h) }
```
```suggestion
  rcases n.eq_zero_or_pos with rfl | hn,
  { exact dvd_zero p },
  { rwa ←mem_factors_iff_dvd hn (prime_of_mem_factors h) }
```

some golf - can you tell I really like `rcases`? I'm surprised your convert thing worked, ngl

ericrbg commented on src/data/nat/prime.lean:
```diff
@@ -976,3 +983,47 @@ namespace nat
+  (p^k).factors.to_finset = {p} :=
+begin
+  rw hp.factors_pow, exact list.to_finset_repeat_of_pos hk,
+end
```
```suggestion
by rw [hp.factors_pow, list.to_finset_repeat_of_pos hk]
```

🏌️

ericrbg commented on src/data/nat/prime.lean:
```diff
@@ -976,3 +983,47 @@ namespace nat
+  p ∈ (a * b).factors ↔ p ∈ a.factors ∨ p ∈ b.factors :=
+begin
+  rw [mem_factors (mul_pos ha hb), mem_factors ha, mem_factors hb, ←and_or_distrib_left],
+  simp only [and.congr_right_iff], exact prime.dvd_mul,
```
```suggestion
  rw [mem_factors (mul_pos ha hb), mem_factors ha, mem_factors hb, ←and_or_distrib_left],
  simpa only [and.congr_right_iff] using prime.dvd_mul,
```

this isn't really a golf but more that mathlib style usually doesn't put more than one tactic per line (unless using the `;` combinator, or a very short `by` block)

ericrbg commented on src/data/nat/prime.lean:
```diff
@@ -976,3 +983,47 @@ namespace nat
+  (a * b).factors.to_finset = a.factors.to_finset ∪ b.factors.to_finset :=
+begin
+  ext p, rw finset.mem_union, repeat {rw list.mem_to_finset}, exact factors_mul_of_pos ha hb p,
+end
```
```suggestion
by { ext p, simp only [finset.mem_union, list.mem_to_finset, factors_mul_of_pos ha hb p] }
```

for example, this is a `by` block so it's more ok. also, `repeat rw` is pretty much always a worse version of `simp only`

ericrbg commented on src/number_theory/divisors.lean:
```diff
@@ -400,4 +400,16 @@ begin
+    { ext q,
+      rw [list.mem_to_finset, mem_filter, mem_divisors, mem_factors (pos_iff_ne_zero.mpr hn)],
+      rw [and_comm, and.congr_left_iff],
+      intros hq, simp only [iff_self_and, ne.def], exact imp_intro hn }
```
```suggestion
    rcases n.eq_zero_or_pos with rfl | hn,
    { simp },
    { ext q,
      simpa [hn, hn.ne', mem_factors] using and_comm (prime q) (q ∣ n) }
```

stuart-presnell replied:
Thanks for this, these are a couple of beyond-absolute-beginner techniques that I hadn't picked up yet.

ericrbg commented on src/data/nat/prime.lean:
```diff
@@ -976,3 +983,47 @@ namespace nat
+  by_cases hb : b = 0, { rw hb at *, simp at hab, rw hab, simp },
+  rw factors_mul_of_pos (pos_iff_ne_zero.mpr ha) (pos_iff_ne_zero.mpr hb) p,
+  rw list.mem_union,
+end
```
```suggestion
begin
  rcases a.eq_zero_or_pos with rfl | ha,
  { simp [(coprime_zero_left _).mp hab] },
  rcases b.eq_zero_or_pos with rfl | hb,
  { simp [(coprime_zero_right _).mp hab] },
  rw [factors_mul_of_pos ha hb p, list.mem_union]
end
```

golfing and removing non-terminal `simp`s

ericrbg reviewed (commented):
golfed your proofs a little. the process is usually just to wait for a maintainer to reply and approve it, but they're busy people

stuart-presnell committed:
Lots of golfing and style improvements per reviewer comments

Vierkantor commented on src/data/finset/basic.lean:
```diff
@@ -1823,6 +1823,10 @@ end
   to_finset l.reverse = l.to_finset :=
 to_finset_eq_of_perm _ _ (reverse_perm l)
 
+lemma to_finset_repeat_of_pos {a : α} {n : ℕ} (hn : n ≠ 0):
```
I'd change the name slightly to be extra precise what the hypothesis is:
```suggestion
lemma to_finset_repeat_of_ne_zero {a : α} {n : ℕ} (hn : n ≠ 0):
```

Vierkantor commented on src/data/nat/prime.lean:
```diff
@@ -976,3 +983,44 @@ namespace nat
+/-- If `a`,`b` are positive the prime divisors of `(a * b)` are the union of those of `a` and `b` -/
+lemma prime_divisors_mul_of_pos {a b : ℕ} (ha : 0 < a) (hb : 0 < b) :
+  (a * b).factors.to_finset = a.factors.to_finset ∪ b.factors.to_finset :=
+by { ext p, simp only [finset.mem_union, list.mem_to_finset, factors_mul_of_pos ha hb p] }
```
```suggestion
/-- If `a`,`b` are positive the prime divisors of `(a * b)` are the union of those of `a` and `b` -/
lemma factors_mul_of_pos {a b : ℕ} (ha : 0 < a) (hb : 0 < b) :
  (a * b).factors.to_finset = a.factors.to_finset ∪ b.factors.to_finset :=
by { ext p, simp only [finset.mem_union, list.mem_to_finset, factors_mul_of_pos ha hb p] }
```

Vierkantor commented on src/number_theory/divisors.lean:
```diff
@@ -400,4 +400,14 @@ begin
 
+/-- The factors of `n` are the prime divisors -/
+lemma prime_divisors_eq_to_filter_divisors_prime (n : ℕ) :
+n.factors.to_finset = (divisors n).filter prime :=
```
```suggestion
  n.factors.to_finset = (divisors n).filter prime :=
```

Vierkantor reviewed (approved):
This is looking quite good, just a few nitpicks left!

bors d+

stuart-presnell committed:
Final tweaks per reviewer suggestions

stuart-presnell committed:
Fixing an error introduced in last commit

Vierkantor commented on src/data/nat/prime.lean:
```diff
@@ -976,3 +983,44 @@ namespace nat
+  (p^k).factors.to_finset = {p} :=
+by rw [hp.factors_pow, list.to_finset_repeat_of_ne_zero hk.ne']
+
+lemma factors_mul_of_pos {a b : ℕ} (ha : 0 < a) (hb : 0 < b) (p : ℕ) :
```
For example, with the following change:
```suggestion
lemma mem_factors_mul_of_pos {a b : ℕ} (ha : 0 < a) (hb : 0 < b) (p : ℕ) :
```

Vierkantor commented on src/data/nat/prime.lean:
```diff
@@ -976,3 +983,44 @@ namespace nat
+end
+
+/-- If `a`,`b` are positive the prime divisors of `(a * b)` are the union of those of `a` and `b` -/
+lemma factors_mul_of_pos {a b : ℕ} (ha : 0 < a) (hb : 0 < b) :
```
Or this one (or both):
```suggestion
lemma to_finset_factors_mul_of_pos {a b : ℕ} (ha : 0 < a) (hb : 0 < b) :
```

stuart-presnell committed:
Fixing a naming clash in `data/nat/prime`
